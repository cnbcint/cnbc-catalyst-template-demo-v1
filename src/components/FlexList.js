import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { breakpoints } from "../helpers/Layout";

const FlexContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-start;

  align-items: inherit;
  margin: -1rem;
  > div {
    width: ${(props) => `calc(${100 / props.columns.mobile}% - 2rem)`};
    margin: 1rem;
    @media (min-width: ${breakpoints.mobile}) {
      width: ${(props) => `calc(${100 / props.columns.tablet}% - 2rem)`};
    }
    @media (min-width: ${breakpoints.tablet}) {
      width: ${(props) => `calc(${100 / props.columns.desktop}% - 2rem)`};
    }
  }
`;

export default function FlexList(props) {
  return (
    <FlexContainer
      columns={props.columns}
      className={props.className}
      style={props.style}
    >
      {props.children}
    </FlexContainer>
  );
}

FlexList.defaultProps = {
  columns: { mobile: 1, desktop: 3, tablet: 2 },
};

FlexList.propTypes = {
  className: PropTypes.string,
  children: PropTypes.array.isRequired,
  style: PropTypes.object,
  columns: PropTypes.shape({
    mobile: PropTypes.number,
    desktop: PropTypes.number,
    tablet: PropTypes.number,
  }),
};
