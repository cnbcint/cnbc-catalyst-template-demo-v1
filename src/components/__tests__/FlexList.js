import React from "react";
import FlexList from "../FlexList";
import { render } from "@testing-library/react";
import "jest-styled-components";
import { css } from "styled-components";

describe("FlexList", () => {
  it("Renders correctly", () => {
    const { container } = render(
      <FlexList>
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </FlexList>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("has the correct media queries for default", () => {
    const { container } = render(
      <FlexList>
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </FlexList>
    );
    expect(container.firstChild).toHaveStyleRule("width", "calc(100% - 2rem)", {
      modifier: css`> div`,
    });
    expect(container.firstChild).toHaveStyleRule("width", "calc(50% - 2rem)", {
      media: `(min-width:768px)`,
      modifier: css`> div`,
    });
    expect(container.firstChild).toHaveStyleRule(
      "width",
      "calc(33.333333333333336% - 2rem)",
      {
        media: `(min-width:1280px)`,
        modifier: css`> div`,
      }
    );
  });
  it("generates the correct media queries based on columns props", () => {
    const { container } = render(
      <FlexList columns={{ mobile: 1, tablet: 2, desktop: 4 }}>
        <div>1</div>
        <div>2</div>
        <div>3</div>
      </FlexList>
    );
    expect(container.firstChild).toHaveStyleRule("width", "calc(100% - 2rem)", {
      modifier: css`> div`,
    });
    expect(container.firstChild).toHaveStyleRule("width", "calc(50% - 2rem)", {
      media: `(min-width:768px)`,
      modifier: css`> div`,
    });
    expect(container.firstChild).toHaveStyleRule("width", "calc(25% - 2rem)", {
      media: `(min-width:1280px)`,
      modifier: css`> div`,
    });
  });
});
