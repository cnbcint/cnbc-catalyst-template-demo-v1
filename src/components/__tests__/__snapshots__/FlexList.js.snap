// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlexList Renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: inherit;
  -webkit-box-align: inherit;
  -ms-flex-align: inherit;
  align-items: inherit;
  margin: -1rem;
}

.c0 > div {
  width: calc(100% - 2rem);
  margin: 1rem;
}

@media (min-width:768px) {
  .c0 > div {
    width: calc(50% - 2rem);
  }
}

@media (min-width:1280px) {
  .c0 > div {
    width: calc(33.333333333333336% - 2rem);
  }
}

<div
  class="c0"
>
  <div>
    1
  </div>
  <div>
    2
  </div>
  <div>
    3
  </div>
</div>
`;
